pkg/loop/internal/relayer/pluginprovider/chainreader/chain_reader.go

```diff
@@ -172,10 +229,249 @@ func (c *Server) Bind(ctx context.Context, bindings *pb.BindRequest) (*emptypb.E
+			return query.And(expressions...), nil
+		}
+		return query.Or(expressions...), nil
+	case *pb.Expression_Primitive:
+		switch primitive := pbEvaluatedExpr.Primitive.GetPrimitive().(type) {
+		case *pb.Primitive_Comparer:
+			var valComparers []query.ValueComparer
+			for _, pbValComparer := range primitive.Comparer.ValueComparers {
+				valComparers = append(valComparers, query.ValueComparer{Value: pbValComparer.Value, Operator: query.ComparisonOperator(pbValComparer.Operator)})
+			}
+			return query.Comparer(primitive.Comparer.Name, valComparers...), nil
+		case *pb.Primitive_Confirmations:
+			return query.Confirmation(query.ConfirmationLevel(primitive.Confirmations.Confirmations)), nil
+		case *pb.Primitive_Block:
+			return query.Block(primitive.Block.BlockNumber, query.ComparisonOperator(primitive.Block.Operator)), nil
+		case *pb.Primitive_TxHash:
+			return query.TxHash(primitive.TxHash.TxHash), nil
+		case *pb.Primitive_Timestamp:
+			return query.Timestamp(primitive.Timestamp.Timestamp, query.ComparisonOperator(primitive.Timestamp.Operator)), nil
+		default:
+			return query.Expression{}, status.Errorf(codes.InvalidArgument, "Unknown expression type")
+		}
+	default:
+		return query.Expression{}, status.Errorf(codes.InvalidArgument, "Unknown expression type")
+	}
+}
+
+func convertLimitAndSortFromProto(limitAndSort *pb.LimitAndSort) (query.LimitAndSort, error) {
+	var sortByArr []query.SortBy
+	for _, sortBy := range limitAndSort.SortBy {
+		switch sort := sortBy.SortBy.(type) {
+		case *pb.SortBy_SortByTimestamp:
+			sortByArr = append(sortByArr, query.NewSortByTimestamp(query.SortDirection(sort.SortByTimestamp.GetSortDirection())))
+		case *pb.SortBy_SortByBlock:
+			sortByArr = append(sortByArr, query.NewSortByBlock(query.SortDirection(sort.SortByBlock.GetSortDirection())))
+		case *pb.SortBy_SortBySequence:
+			sortByArr = append(sortByArr, query.NewSortBySequence(query.SortDirection(sort.SortBySequence.GetSortDirection())))
+		default:
+			return query.LimitAndSort{}, status.Errorf(codes.InvalidArgument, "Unknown order by type")
+		}
+	}
+
+	limit := limitAndSort.Limit
+	cursorDefined := limit.Cursor != nil
+	cursorDirectionDefined := limit.Direction != nil
+	if cursorDefined && cursorDirectionDefined {
+		return query.NewLimitAndSort(query.CursorLimit(*limit.Cursor, (query.CursorDirection)(*limit.Direction), limit.Count)), nil
+	} else if (!cursorDefined && cursorDirectionDefined) || (cursorDefined && !cursorDirectionDefined) {
+		return query.LimitAndSort{}, status.Errorf(codes.InvalidArgument, "Limit cursor and cursor direction must both be defined or undefined")
+	}
+
+	return query.NewLimitAndSort(query.CountLimit(limit.Count), sortByArr...), nil
+}
+
+func convertSequencesFromProto(pbSequences []*pb.Sequence, sequenceDataType any) ([]types.Sequence, error) {
+	var sequences []types.Sequence
+	for _, pbSequence := range pbSequences {
+		cpy := reflect.New(reflect.TypeOf(sequenceDataType).Elem()).Interface()
```

Comment:
We can also check if sequenceDataType is a non pointer and then just do reflect.New(sequenceDataType). I don't think that changes performance that much though. I ran some benchmark and this doesn't impact performance too much which was surprising.

Reply:
Because we don't actually write to the `sequenceDataType` and only use it for typing future elements, it doesn't really matter if it comes in a pointer or not. I would say leave it alone but if we do normalize, normalize to a non-pointer.